fortls/parse_fortran.py

```diff
@@ -1124,6 +1212,657 @@ def check_file(self, obj_tree, max_line_length=-1, max_comment_line_length=-1):
+                file_ast.add_scope(new_enum, FRegex.END_ENUMD, req_container=True)
+                self.parser_debug("ENUM", line, line_number)
+
+            elif obj_type == "int":
+                name = obj_info.name
+                if name is None:
+                    counters["interface"] += 1
+                    name = f"#GEN_INT{counters['interface']}"
+                new_int = fortran_int(
+                    file_ast, line_number, name, abstract=obj_info.abstract
+                )
+                file_ast.add_scope(new_int, FRegex.END_INT, req_container=True)
+                self.parser_debug("INTERFACE", line, line_number)
+
+            elif obj_type == "gen":
+                new_int = fortran_int(
+                    file_ast, line_number, obj_info.bound_name, abstract=False
+                )
+                new_int.set_visibility(obj_info.vis_flag)
+                file_ast.add_scope(new_int, FRegex.END_INT, req_container=True)
+                for pro_link in obj_info.pro_links:
+                    file_ast.add_int_member(pro_link)
+                file_ast.end_scope(line_number)
+                self.parser_debug("GENERIC", line, line_number)
+
+            elif obj_type == "int_pro":
+                if file_ast.current_scope is not None:
+                    if file_ast.current_scope.get_type() == INTERFACE_TYPE_ID:
+                        for name in obj_info:
+                            file_ast.add_int_member(name)
+                        self.parser_debug("INTERFACE-PRO", line, line_number)
+
+                    elif file_ast.current_scope.get_type() == SUBMODULE_TYPE_ID:
+                        new_impl = fortran_scope(file_ast, line_number, obj_info[0])
+                        file_ast.add_scope(new_impl, FRegex.END_PRO)
+                        self.parser_debug("INTERFACE_IMPL", line, line_number)
+
+            elif obj_type == "use":
+                file_ast.add_use(
+                    obj_info.mod_name,
+                    line_number,
+                    obj_info.only_list,
+                    obj_info.rename_map,
+                )
+                self.parser_debug("USE", line, line_number)
+
+            elif obj_type == "import":
+                file_ast.add_use("#IMPORT", line_number, obj_info)
+                self.parser_debug("IMPORT", line, line_number)
+
+            elif obj_type == "inc":
+                file_ast.add_include(obj_info, line_number)
+                self.parser_debug("INCLUDE", line, line_number)
+
+            elif obj_type == "vis":
+                if file_ast.current_scope is None:
+                    file_ast.parse_errors.append(
+                        {
+                            "line": line_number,
+                            "schar": 0,
+                            "echar": 0,
+                            "mess": "Visibility statement without enclosing scope",
+                            "sev": 1,
+                        }
```

Comment:
Encapsulate, the interface for parse_errors does not need to be this. It is dictated by the diagnostic fields